File: code_cpp/BST/11_Top_view_of_binary_tree.cpp
/*  


compile:
    g++ -std=c++0x 10_vertical_traversal_of_binary_tree.cpp -o 10_vertical_traversal_of_binary_tree
run:
    ./10_vertical_traversal_of_binary_tree  

*/
#include<bits/stdc++.h>
using namespace std;
// Constructing Binary Tree
// create node of tree 
struct Node{

    int key;
    Node *left;
    Node *right;
    Node(int k){
        key=k;
        left=NULL;
        right=NULL;
    }
};

// Top view of binary tree
void vTraversal(Node *root){
    map<int,vector<int> > mp;
    queue<pair<Node*,int> > q;
    q.push({root,0});
    while(q.empty()==false){
        auto p=q.front();
        Node *curr=p.first;
        int hd=p.second;
        if(mp.find(hd)==mp.end())
            mp[hd].push_back(curr->key);
        q.pop();
        if(curr->left!=NULL)
            q.push({curr->left,hd-1});
        if(curr->right!=NULL)
            q.push({curr->right,hd+1});
    }
    // print value of map
    for(auto x:mp){
        vector<int> v=x.second;
        for(int y:v)
            cout<<y<<" ";
    }
}



// print elements
vector<int>v;
void inorder(Node *root){
    if(root!=NULL){
        inorder(root->left);
        cout<<root->key<<" ";
        v.push_back(root->key);
        inorder(root->right);    
    }
} 
int main(){
    // Node *root=NULL;    // empty tree
    Node *root=new Node(15);
    root->left=new Node(5);
    root->right=new Node(20);
    root->left->left=new Node(3);
    root->right->right=new Node(80);
    root->right->left=new Node(18);
    root->right->left->left=new Node(16);

    
    vTraversal(root);
    // vSumMain(root);
    // inorder(root);
   
    return 0;
}

File: code_java/self pased DSA/BST/11_Top_view_of_binary_tree.java
/**
 * Problem: Print the top view of a binary tree.
 *
 * Definition:
 * The top view of a binary tree is the set of nodes visible when the tree is viewed from the top.
 * For each vertical line (i.e., horizontal distance from root), only the first node encountered in top-down level order traversal is included.
 *
 * Approach:
 * - We perform a level order traversal using a queue and track horizontal distances (hd).
 * - We use a TreeMap to store the first node for each horizontal distance.
 * - If a horizontal distance is encountered for the first time, we add the node's value to the map.
 * - Finally, we print the values in order of increasing horizontal distance (from left to right).
 *
 * Why this approach:
 * - Level order traversal ensures we visit the top-most node first for each vertical column.
 * - TreeMap keeps the keys sorted, so we get the left-to-right top view directly.
 *
 * This approach works because it captures the first node seen from top-down at every horizontal distance efficiently.
 * Time Complexity: O(n log n) due to TreeMap operations.
 */

import java.util.*;

class Node {
    int key;
    Node left, right;

    Node(int key) {
        this.key = key;
        left = right = null;
    }
}

public class TopViewBinaryTree {

    static class Pair {
        Node node;
        int hd;

        Pair(Node node, int hd) {
            this.node = node;
            this.hd = hd;
        }
    }

    public static void topView(Node root) {
        if (root == null) return;

        TreeMap<Integer, Integer> map = new TreeMap<>();
        Queue<Pair> q = new LinkedList<>();
        q.add(new Pair(root, 0));

        while (!q.isEmpty()) {
            Pair p = q.poll();
            Node curr = p.node;
            int hd = p.hd;

            map.putIfAbsent(hd, curr.key);

            if (curr.left != null) q.add(new Pair(curr.left, hd - 1));
            if (curr.right != null) q.add(new Pair(curr.right, hd + 1));
        }

        for (int val : map.values()) {
            System.out.print(val + " ");
        }
        System.out.println();
    }

    public static void main(String[] args) {
        Node root = new Node(15);
        root.left = new Node(5);
        root.right = new Node(20);
        root.left.left = new Node(3);
        root.right.right = new Node(80);
        root.right.left = new Node(18);
        root.right.left.left = new Node(16);

        topView(root);
    }
}
